### Finding Minimum Path in an image array

##### By finding the minimum path in image resizing you'll find the pixels that you can cut without losing quality on the image

In [77]:
import numpy as np

# Example Array

In [78]:
img = np.array([[0.1,0.8,0.8,0.3,0.5,0.4],
                [0.7,0.8,0.1,0.0,0.8,0.4],
                [0.8,0.0,0.4,0.7,0.2,0.9],
                [0.9,0.0,0.0,0.5,0.9,0.4],
                [0.2,0.4,0.0,0.2,0.4,0.5],
                [0.2,0.4,0.2,0.5,0.3,0.0]])

# Finding Minimum Path 
<b> Not efficient like this because you may not find the true minimum path </b>

In [79]:
base = img[0][3]
debut = 2
end = 5

A = [base]
for i in range(1,len(img)):
    values = img[i][debut:end]

    A.append(min(values))
    index_min = min(range(len(values)), key=values.__getitem__)
    if index_min == 0:
        debut -= 1
        end -= 1
    if index_min == 2:
        debut += 1
        end += 1

A,sum(A)

([0.3, 0.0, 0.2, 0.4, 0.4, 0.0], 1.3)

# Minimal Energy to bottom

output = np.zeros(img.shape)

In [80]:
output = np.zeros(img.shape)
output[-1] = img[-1]
array = []
for i in range(1,len(output)):
    for j in range(len(output)):
        if j == 0:
            array = output[len(output)-i][j:j+2]

        elif j == len(output)-1:
            array = output[len(output)-i][j-1:j+1]

        else:
            array = output[len(output)-i][j-1:j+2]

        minArray = min(array)
        output[len(output)-i-1][j] = img[len(output)-i-1][j] + minArray

output

array([[1. , 1.1, 1.1, 0.6, 1.1, 1.7],
       [0.9, 1. , 0.3, 0.6, 1.7, 1.3],
       [1. , 0.2, 0.6, 0.9, 0.9, 1.7],
       [1.3, 0.2, 0.2, 0.7, 1.3, 0.8],
       [0.4, 0.6, 0.2, 0.4, 0.4, 0.5],
       [0.2, 0.4, 0.2, 0.5, 0.3, 0. ]])

# Finding Minimum Path 
<b> After calculating the minimum energy  efficient way </b>

In [81]:
index_min = min(range(len(output)), key=output[0].__getitem__)
base = img[0][index_min]

# you can do this with less code but it's just to better understand 
if index_min == 0:
    debut = index_min 
    end = index_min + 2 

if index_min > 0:
    debut = index_min - 1
    end = index_min + 2 

if index_min == len(output)-1:
     debut = index_min - 2
     end = index_min


A = [base]
for i in range(1,len(img)):

    values = output[i][debut:end]

    index_min = min(range(len(values)), key=values.__getitem__)

    A.append(img[i][debut+index_min])
    if index_min == 0:
        debut -= 1
        end -= 1
    if index_min == 2:
        debut += 1
        end += 1

A,sum(A)

([0.3, 0.1, 0.0, 0.0, 0.0, 0.2], 0.6000000000000001)

### We got 1.3 as the sum of the minimum path from the first algorithm then after doing the minimum energy we can see that the sum of the true minimum path is 0.6